cpp/dolfin/common/IndexMap.cpp

```diff
@@ -1,113 +1,137 @@
-  std::vector<std::int32_t> displs_forward(num_neighbours + 1, 0);
-  for (std::int32_t i = 0; i < num_neighbours; ++i)
+  // Check for 'symmetry' of the graph
+#if DEBUG
   {
-    displs_reverse[i + 1] = displs_reverse[i] + reverse_sizes[i];
-    displs_forward[i + 1] = displs_forward[i] + _forward_sizes[i];
+    std::vector<int> sources(num_neighbours), dests(num_neighbours);
+    MPI_Dist_graph_neighbors(_neighbour_comm, num_neighbours, sources.data(),
+                             NULL, num_neighbours, dests.data(), NULL);
+    assert(sources == dests);
+    assert(sources == neighbours);
   }
+#endif
 
-  // Create vectors for forward communication
-  int n_send = displs_reverse.back();
-  std::vector<std::int32_t> reverse_indices(n_send);
-  int n_recv = displs_forward.back();
-  _forward_indices.resize(n_recv);
-
-  std::vector<std::int32_t> displs(displs_reverse);
-  for (std::int32_t j = 0; j < _ghosts.size(); ++j)
+  // Create displacement vectors
+  std::vector<int> disp_out(num_neighbours + 1, 0),
+      disp_in(num_neighbours + 1, 0);
+  std::partial_sum(out_edges_num.begin(), out_edges_num.end(),
+                   disp_out.begin() + 1);
+  std::partial_sum(in_edges_num.begin(), in_edges_num.end(),
+                   disp_in.begin() + 1);
+
+  // Get rank on neighbourhood communicator for each ghost, and for each
+  // ghost compute the local index on the owning process
+  std::vector<int> out_indices(disp_out.back());
+  std::vector<int> disp(disp_out);
+  for (int j = 0; j < _ghosts.size(); ++j)
   {
-    const int p = _ghost_owners[j];
-    const auto it = std::find(_neighbours.begin(), _neighbours.end(), p);
-    assert(it != _neighbours.end());
-    const int np = it - _neighbours.begin();
+    // Get rank of owner process rank on global communicator
+    const int p = ghost_owner_global[j];
 
-    // Convert to neighbour number instead of global process number
+    // Get rank of owner on neighbourhood communicator
+    const auto it = std::find(neighbours.begin(), neighbours.end(), p);
+    assert(it != neighbours.end());
+    const int np = std::distance(neighbours.begin(), it);
+
+    // Store owner neighbourhood rank for each ghost
     _ghost_owners[j] = np;
 
-    reverse_indices[displs[np]] = _ghosts[j] - _all_ranges[p];
-    ++displs[np];
+    // Local on owning process
+    out_indices[disp[np]] = _ghosts[j] - _all_ranges[p];
+    disp[np] += 1;
   }
 
   //  May have repeated shared indices with different processes
-  MPI_Neighbor_alltoallv(reverse_indices.data(), reverse_sizes.data(),
-                         displs_reverse.data(), MPI_INT,
-                         _forward_indices.data(), _forward_sizes.data(),
-                         displs_forward.data(), MPI_INT, _neighbour_comm);
+  std::vector<int> indices_in(disp_in.back());
+  MPI_Neighbor_alltoallv(
+      out_indices.data(), out_edges_num.data(), disp_out.data(), MPI_INT,
+      indices_in.data(), // out
```

Comment:
What does the comment "out" mean in this position?

Reply:
It follows the MPI and graph jargon for edges, defining 'out' edges as ghost to owner (chosen as this is what we have to start), with 'in' edges being owner -> ghost. I think it's clearer thinking of it as a graph in terms of forward and reverse.

I think eventually we should store two neighbourhood communicators for communication in different directions.